File: merge_words/python/gui.py
#!/usr/bin/env python3

import asyncio
import platform
import pygame
import numpy as np
from merge_words import MergeWords

FPS = 60
TILE_SIZE = (80, 100)
BUTTON_HEIGHT = 50
FONT_SIZE = 24

async def main():
    pygame.init()
    screen = pygame.display.set_mode((300, 600))
    pygame.display.set_caption("Merge Words")
    font = pygame.font.SysFont('arial', FONT_SIZE)
    game = MergeWords()
    is_dark_theme = False

    # Sound effects
    def play_sound(frequency, duration=0.1):
        sample_rate = 44100
        t = np.linspace(0, duration, int(sample_rate * duration), False)
        wave = np.sin(2 * np.pi * frequency * t) * 0.5
        stereo = np.vstack((wave, wave)).T.astype(np.float32)
        sound = pygame.sndarray.make_sound(stereo)
        sound.play()

    # Colors
    def get_colors():
        return {
            'bg': (237, 242, 247) if not is_dark_theme else (26, 32, 44),
            'tile': (255, 255, 255) if not is_dark_theme else (45, 55, 72),
            'tile_border': (160, 174, 192) if not is_dark_theme else (74, 85, 104),
            'tile_selected': (104, 211, 145) if not is_dark_theme else (56, 161, 105),
            'text': (45, 55, 72) if not is_dark_theme else (226, 232, 240),
            'button': (34, 197, 94) if not is_dark_theme else (52, 211, 153),
            'button_hover': (22, 163, 74) if not is_dark_theme else (44, 181, 130),
            'rules_bg': (226, 232, 240) if not is_dark_theme else (45, 55, 72)
        }

    # Buttons
    buttons = [
        {'text': 'Submit', 'rect': pygame.Rect(10, 350, 80, BUTTON_HEIGHT), 'action': lambda: game.submit_word(), 'sound': 500},
        {'text': 'Shuffle', 'rect': pygame.Rect(110, 350, 80, BUTTON_HEIGHT), 'action': game.shuffle_tiles, 'sound': 400},
        {'text': 'Backspace', 'rect': pygame.Rect(210, 350, 80, BUTTON_HEIGHT), 'action': game.backspace, 'sound': 300},
        {'text': 'Reset', 'rect': pygame.Rect(10, 10, 80, BUTTON_HEIGHT), 'action': game.initialize_game, 'sound': 600},
        {'text': 'Theme', 'rect': pygame.Rect(210, 10, 80, BUTTON_HEIGHT), 'action': lambda: toggle_theme(), 'sound': 350}
    ]

    def toggle_theme():
        nonlocal is_dark_theme
        is_dark_theme = not is_dark_theme

    def draw():
        colors = get_colors()
        screen.fill(colors['bg'])

        # Rules
        rules_rect = pygame.draw.rect(screen, colors['rules_bg'], (10, 60, 280, 100), border_radius=5)
        rules_text = [
            "Merge Words Rules:",
            "1. Click tiles to form a word",
            "2. Submit or Enter to check",
            "3. Score = word length",
            "4. Shuffle, backspace, or reset"
        ]
        for i, line in enumerate(rules_text):
            screen.blit(font.render(line, True, colors['text']), (15, 65 + i * 20))

        # Score and Current Word
        screen.blit(font.render(f"Score: {game.score}", True, colors['text']), (10, 170))
        screen.blit(font.render(f"Word: {game.get_current_word()}", True, colors['text']), (10, 200))

        # Grid
        for i in range(9):
            x, y = (i % 3) * TILE_SIZE[0] + 30, (i // 3) * TILE_SIZE[1] + 230
            color = colors['tile_selected'] if i in game.selected_tiles else colors['tile']
            pygame.draw.rect(screen, color, (x, y, TILE_SIZE[0], TILE_SIZE[1]), border_radius=5)
            pygame.draw.rect(screen, colors['tile_border'], (x, y, TILE_SIZE[0], TILE_SIZE[1]), 2, border_radius=5)
            text = font.render(''.join(game.tiles[i]), True, colors['text'])
            text_rect = text.get_rect(center=(x + TILE_SIZE[0] // 2, y + TILE_SIZE[1] // 2))
            screen.blit(text, text_rect)

        # Buttons
        for button in buttons:
            color = colors['button_hover'] if button['rect'].collidepoint(pygame.mouse.get_pos()) else colors['button']
            pygame.draw.rect(screen, color, button['rect'], border_radius=5)
            text = font.render(button['text'], True, colors['text'])
            text_rect = text.get_rect(center=button['rect'].center)
            screen.blit(text, text_rect)

        # Found Words
        pygame.draw.rect(screen, colors['rules_bg'], (10, 410, 280, 180), border_radius=5)
        screen.blit(font.render("Found Words:", True, colors['text']), (15, 415))
        for i, word in enumerate(game.found_words[:8]):
            screen.blit(font.render(word, True, colors['text']), (15, 435 + i * 20))

        # Game Over
        if game.is_game_over():
            overlay = pygame.Surface((300, 600))
            overlay.fill((0, 0, 0))
            overlay.set_alpha(128)
            screen.blit(overlay, (0, 0))
            pygame.draw.rect(screen, colors['rules_bg'], (50, 200, 200, 200), border_radius=10)
            screen.blit(font.render("Game Over!", True, colors['text']), (100, 250))
            screen.blit(font.render(f"Score: {game.score}", True, colors['text']), (100, 280))
            play_again_rect = pygame.Rect(100, 320, 100, 40)
            color = colors['button_hover'] if play_again_rect.collidepoint(pygame.mouse.get_pos()) else colors['button']
            pygame.draw.rect(screen, color, play_again_rect, border_radius=5)
            screen.blit(font.render("Play Again", True, colors['text']), (110, 330))

        pygame.display.flip()

    def handle_click(pos):
        for button in buttons:
            if button['rect'].collidepoint(pos):
                play_sound(button['sound'])
                button['action']()
                return
        if game.is_game_over():
            if pygame.Rect(100, 320, 100, 40).collidepoint(pos):
                play_sound(600)
                game.initialize_game()
            return
        for i in range(9):
            x, y = (i % 3) * TILE_SIZE[0] + 30, (i // 3) * TILE_SIZE[1] + 230
            if pygame.Rect(x, y, TILE_SIZE[0], TILE_SIZE[1]).collidepoint(pos):
                game.select_tile(i)
                return

    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            elif event.type == pygame.MOUSEBUTTONDOWN:
                handle_click(event.pos)
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_RETURN:
                    if game.is_game_over():
                        play_sound(600)
                        game.initialize_game()
                    else:
                        play_sound(500)
                        game.submit_word()
                elif event.key == pygame.K_BACKSPACE:
                    play_sound(300)
                    game.backspace()

        draw()
        await asyncio.sleep(1.0 / FPS)

    pygame.quit()

if platform.system() == "Emscripten":
    asyncio.ensure_future(main())
else:
    if __name__ == "__main__":
        asyncio.run(main())

File: merge_words/java/GUI.java
package merge_words.java;

import javax.sound.sampled.*;
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;

public class GUI extends JFrame {
    private final MergeWords game;
    private boolean isDarkTheme = false;
    private final JLabel scoreLabel, wordLabel;
    private final JPanel gridPanel, buttonPanel, wordListPanel;
    private final JDialog gameOverDialog;
    private Color bgColor, tileColor, tileSelectedColor, tileBorderColor, textColor, buttonColor, buttonHoverColor, rulesBgColor;

    public GUI() {
        game = new MergeWords();
        setTitle("Merge Words");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(350, 650);
        setLocationRelativeTo(null);
        setLayout(new BorderLayout(10, 10));
        updateTheme();

        // Rules
        JPanel rulesPanel = new JPanel();
        rulesPanel.setBackground(rulesBgColor);
        rulesPanel.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));
        rulesPanel.setLayout(new BoxLayout(rulesPanel, BoxLayout.Y_AXIS));
        String[] rules = {
            "Merge Words Rules:", "1. Click tiles to form a word", "2. Submit or Enter to check",
            "3. Score = word length", "4. Shuffle, backspace, or reset"
        };
        for (String rule : rules) {
            JLabel label = new JLabel(rule);
            label.setForeground(textColor);
            rulesPanel.add(label);
        }
        add(rulesPanel, BorderLayout.NORTH);

        // Top Buttons
        JPanel topPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 10, 5));
        topPanel.setBackground(bgColor);
        JButton resetButton = createButton("Reset", _ -> { playSound(600); game.initializeGame(); updateUI(); });
        JButton themeButton = createButton("Theme", _ -> { playSound(350); isDarkTheme = !isDarkTheme; updateTheme(); updateUI(); });
        topPanel.add(resetButton);
        topPanel.add(themeButton);
        add(topPanel, BorderLayout.NORTH);
        rulesPanel.setBorder(BorderFactory.createEmptyBorder(0, 10, 10, 10));

        // Center Panel
        JPanel centerPanel = new JPanel();
        centerPanel.setBackground(bgColor);
        centerPanel.setLayout(new BoxLayout(centerPanel, BoxLayout.Y_AXIS));

        scoreLabel = new JLabel("Score: 0");
        scoreLabel.setForeground(textColor);
        scoreLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
        centerPanel.add(scoreLabel);

        wordLabel = new JLabel("Word: ");
        wordLabel.setForeground(textColor);
        wordLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
        centerPanel.add(wordLabel);

        gridPanel = new JPanel(new GridLayout(3, 3, 0, 0));
        gridPanel.setBackground(bgColor);
        gridPanel.setMaximumSize(new Dimension(240, 300));
        centerPanel.add(gridPanel);

        buttonPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 10, 5));
        buttonPanel.setBackground(bgColor);
        buttonPanel.add(createButton("Submit", _ -> { playSound(500); String msg = game.submitWord(); if (!msg.startsWith("Valid")) JOptionPane.showMessageDialog(this, msg); updateUI(); }));
        buttonPanel.add(createButton("Shuffle", _ -> { playSound(400); game.shuffleTiles(); updateUI(); }));
        buttonPanel.add(createButton("Backspace", _ -> { playSound(300); game.backspace(); updateUI(); }));
        centerPanel.add(buttonPanel);

        add(centerPanel, BorderLayout.CENTER);

        // Word List
        wordListPanel = new JPanel();
        wordListPanel.setBackground(rulesBgColor);
        wordListPanel.setLayout(new BoxLayout(wordListPanel, BoxLayout.Y_AXIS));
        wordListPanel.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));
        updateWordList();
        add(wordListPanel, BorderLayout.SOUTH);

        // Keyboard Support
        addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (e.getKeyCode() == KeyEvent.VK_ENTER) {
                    if (game.isGameOver()) {
                        playSound(600);
                        gameOverDialog.dispose();
                        game.initializeGame();
                        updateUI();
                    } else {
                        playSound(500);
                        String msg = game.submitWord();
                        if (!msg.startsWith("Valid")) JOptionPane.showMessageDialog(GUI.this, msg);
                        updateUI();
                    }
                } else if (e.getKeyCode() == KeyEvent.VK_BACK_SPACE) {
                    playSound(300);
                    game.backspace();
                    updateUI();
                }
            }
        });
        setFocusable(true);

        // Game Over Dialog
        gameOverDialog = new JDialog(this, "Game Over", true);
        gameOverDialog.setSize(200, 150);
        gameOverDialog.setLocationRelativeTo(this);
        gameOverDialog.setLayout(new BorderLayout());
        JLabel gameOverLabel = new JLabel("", SwingConstants.CENTER);
        gameOverLabel.setForeground(textColor);
        gameOverDialog.add(gameOverLabel, BorderLayout.CENTER);
        JButton playAgainButton = createButton("Play Again", _ -> { playSound(600); gameOverDialog.dispose(); game.initializeGame(); updateUI(); });
        gameOverDialog.add(playAgainButton, BorderLayout.SOUTH);

        updateUI();
    }

    private JButton createButton(String text, ActionListener action) {
        JButton button = new JButton(text);
        button.setBackground(buttonColor);
        button.setForeground(textColor);
        button.setFocusPainted(false);
        button.setBorder(BorderFactory.createLineBorder(textColor, 1));
        button.addActionListener(action);
        button.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseEntered(MouseEvent e) {
                button.setBackground(buttonHoverColor);
            }
            @Override
            public void mouseExited(MouseEvent e) {
                button.setBackground(buttonColor);
            }
        });
        return button;
    }

    private void updateTheme() {
        if (isDarkTheme) {
            bgColor = new Color(26, 32, 44);
            tileColor = new Color(45, 55, 72);
            tileSelectedColor = new Color(56, 161, 105);
            tileBorderColor = new Color(74, 85, 104);
            textColor = new Color(226, 232, 240);
            buttonColor = new Color(52, 211, 153);
            buttonHoverColor = new Color(44, 181, 130);
            rulesBgColor = new Color(45, 55, 72);
        } else {
            bgColor = new Color(237, 242, 247);
            tileColor = new Color(255, 255, 255);
            tileSelectedColor = new Color(104, 211, 145);
            tileBorderColor = new Color(160, 174, 192);
            textColor = new Color(45, 55, 72);
            buttonColor = new Color(34, 197, 94);
            buttonHoverColor = new Color(22, 163, 74);
            rulesBgColor = new Color(226, 232, 240);
        }
    }

    private void updateUI() {
        scoreLabel.setText("Score: " + game.getScore());
        scoreLabel.setForeground(textColor);
        wordLabel.setText("Word: " + game.getCurrentWord());
        wordLabel.setForeground(textColor);
        gridPanel.removeAll();
        for (int i = 0; i < 9; i++) {
            JPanel tile = new JPanel();
            tile.setBackground(game.getSelectedTiles().contains(i) ? tileSelectedColor : tileColor);
            tile.setBorder(BorderFactory.createLineBorder(tileBorderColor, 2));
            tile.setPreferredSize(new Dimension(80, 100));
            JLabel label = new JLabel(String.join("", game.getTiles().get(i)), SwingConstants.CENTER);
            label.setForeground(textColor);
            tile.add(label);
            int index = i;
            tile.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    if (!game.isGameOver()) {
                        game.selectTile(index);
                        updateUI();
                    }
                }
            });
            gridPanel.add(tile);
        }
        updateWordList();
        if (game.isGameOver()) {
            gameOverDialog.getContentPane().setBackground(rulesBgColor);
            ((JLabel)gameOverDialog.getContentPane().getComponent(0)).setText("<html><center>Game Over!<br>Score: " + game.getScore() + "</center></html>");
            gameOverDialog.getContentPane().getComponent(0).setForeground(textColor);
            gameOverDialog.setVisible(true);
        }
        getContentPane().setBackground(bgColor);
        gridPanel.setBackground(bgColor);
        buttonPanel.setBackground(bgColor);
        revalidate();
        repaint();
    }

    private void updateWordList() {
        wordListPanel.removeAll();
        JLabel title = new JLabel("Found Words:");
        title.setForeground(textColor);
        wordListPanel.add(title);
        for (String word : game.getFoundWords()) {
            JLabel label = new JLabel(word);
            label.setForeground(textColor);
            wordListPanel.add(label);
        }
        wordListPanel.revalidate();
        wordListPanel.repaint();
    }

    private void playSound(int frequency) {
        try {
            AudioFormat format = new AudioFormat(44100, 16, 1, true, false);
            SourceDataLine line = AudioSystem.getSourceDataLine(format);
            line.open(format);
            line.start();
            byte[] buffer = new byte[4410]; // 0.1 seconds
            for (int i = 0; i < buffer.length; i += 2) {
                double angle = i / 44100.0 * frequency * 2.0 * Math.PI;
                short sample = (short)(Math.sin(angle) * 32767 * 0.5);
                buffer[i] = (byte)(sample & 0xFF);
                buffer[i + 1] = (byte)((sample >> 8) & 0xFF);
            }
            line.write(buffer, 0, buffer.length);
            line.drain();
            line.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> new GUI().setVisible(true));
    }
}